### ISSUE TITLE
Using MSW with Next.js causes stopping HMR/Fast Refresh

### ISSUE THREAD

**Issue by tkamenoko** [mswjs/msw#642]
## Describe the bug
Using MSW with Next.js causes stopping HMR/Fast Refresh because of unclosed SSE connections.

## Environment

- `msw: 0.26.2`
- `nodejs: 14.14.0`
- `yarn: 2.4.1`
- `Browser: Chromium 89.0.4389.82`

## To Reproduce
1. Clone https://github.com/tkamenoko/msw-nextjs-hmr .
1. Install deps: `yarn` .
1. Start dev server: `yarn dev` .
1. Open `localhost:3000` in your browser.
1. Show `Network` tab in DevTools.
1. Click rendered links, and you will find that SSE connections between MSW and the dev server are not closed.
1. After the number of connections reaches the limit, HMR will not work.
1. Reload the page, but it will not complete: `Waiting for available sockets`

## Expected behavior
A browser should receive changes and refresh the page.
SSE connections between MSW and the dev server should be closed when connections between browser and MSW are closed.

**Comment by kettanaito** [comment]
Hey. Please update to the latest version of MSW and see if the issue is still there. Thanks.

**Comment by tkamenoko** [comment]
Updating v0.27.1 causes another error. It seem to be fixed by `@msw/cookies` v0.1.4 .
 
![Screenshot_20210312_175837](https://user-images.githubusercontent.com/41364327/110917202-fd031680-835c-11eb-96a0-102ec1b7b223.png)

**Comment by tkamenoko** [comment]
@kettanaito Should I open another issue to talk about this error?

**Comment by kettanaito** [comment]
Hey, @tkamenoko. Thanks for the follow-up. No need to raise an issue, this will be fixed with #649.

**Comment by saintwinkle** [comment]
> Hey, @tkamenoko. Thanks for the follow-up. No need to raise an issue, this will be fixed with #649.

Hello, sorry to bother. This issue hasn't been fixed, I'm suffering the same issue as above.

And not noly React fast refresh, other tools using HMR may have this issue too (Nuxt.js for sure). Better to reopen :)

- `msw: 0.27.1`
- `node: 14.16.0`
- `npm: 7.6.3`
- `browser: Chrome 89.0.4389.90`

**Comment by kettanaito** [comment]
@saintwinkle you're most likely experiencing something else. I suggest opening a new issue and providing a reproduction repository to look at. Thanks.

**Comment by tkamenoko** [comment]
@kettanaito Thanks for updating! Cookies error has been fixed on v0.27.2, but unfortunately Fast Refresh error still exists. 
This issue should be reopened.

**Comment by kettanaito** [comment]
@tkamenoko could you please share more info about the error that still exists? I doubt it's the same error, as the one above was addressed in the cookies package. That's why I suggest treating it as a new issue because it is.

**Comment by saintwinkle** [comment]
@kettanaito I think what @tkamenoko and I mean is that this issue is tracking HMR issue, not the solved cookies issue which maybe not related to this one.😂 So we can still discuss here.

I've created a simpler repo to reproduce the issue: https://github.com/saintwinkle/with-msw-next-app .

And here's the process:

https://user-images.githubusercontent.com/4580163/112440392-3720df00-8d85-11eb-9646-9b8df73879ef.mov


If i'm missing something, let me know. Thanks!

**Comment by KnisterPeter** [comment]
~~Hi, I've discovered this today as well. The serviceworker does seem to swallow updates from the even-source.
With active sw no updates are send to the browser (while the server is working). Without sw the updates are there and it does work fine.
(For me this is not next.js but webpack with webpack-hot-middleware)~~

No, it does work now for me. There must have been another issue in my code.

**Comment by panzelva** [comment]
I have same problem with MSW + Next.js. Can we please reopen this issue?

**Comment by jcheese1** [comment]
Getting the same problem as well.

**Comment by tamj0rd2** [comment]
@KnisterPeter Any idea which changes fixed it for you? Having the same issue shown in the video

**Comment by KnisterPeter** [comment]
@tamj0rd2  We have a quite custom implementation of HMR which was buggy. It had nothing to do with msw.

**Comment by khw1031** [comment]
Having the same Issue
- msw v0.28.2
- next v10.1.2

**Comment by topaxi** [comment]
Still encountering this with msw `0.29.0` and next `10.2.3`.

**Comment by dasblitz** [comment]
Me too, same setup as @topaxi `msw 0.29.0` and `nextjs 10.2.3`. We have a Cypress e2e test in which we test the correct display of pagination buttons (among others). After the sixth navigation trigger Chrome hangs because it starts to infinitely wait for a hmr bundle. If we disable `msw` the test is running fine no matter how often we run it.

I'd be willing to share access to our repository if that would help.

Edit:

I found that if I unregistered the mockServiceWorker Cypress continued to run again. I managed to make my e2e tests more stable by deregistering the mockServiceWorker before each navigation action, like mentioned here:

https://github.com/cypress-io/cypress/issues/702#issuecomment-435873135

Will see if something similar can be put in my next setup during development. But that would ultimately still be a work around. Would really like to see a fix for this, because I really like using msw otherwise!

Edit 2:

I still don't really understand the issue, but we had a product listing page with links to products opening in a new tab. As soon as I had a few 3-4 tabs with our nextjs app open during development one of the tabs would often end up in a page refresh loop. Causing errors like `EventEmitter limit reached` in NodeJS, and `waiting for available Chrome socket` etc.. 

I can 'reproduce' the page refresh loop by hitting the refresh button of the browser multiple times in a row really fast.
